compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,292 @@
+
+  // Verify that the forall has the required GPU thread mapping.
+  if (failed(verifyThreadMapping(forallOp))) {
+    return failure();
+  }
+
+  if (failed(verifyMemoryLayout(dmaOp, rewriter))) {
+    return failure();
+  }
+
+  return success();
+}
+
+static bool matchesTargetDmaSize(int64_t transferSizeBytes,
+                                 ArrayRef<int64_t> targetDmaSizes) {
+  for (int64_t dmaSize : targetDmaSizes) {
+    if (transferSizeBytes == dmaSize) {
+      return true;
+    }
+  }
+  return false;
+}
+
+struct LowerCoalescedGatherDMAPattern
+    : public OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LowerCoalescedGatherDMAPattern(MLIRContext *context,
+                                 ArrayRef<int64_t> targetDmaSizes)
+      : OpRewritePattern<IREE::GPU::CoalescedGatherDMAOp>(context),
+        targetDmaSizes(targetDmaSizes) {}
+
+  LogicalResult matchAndRewrite(IREE::GPU::CoalescedGatherDMAOp dmaOp,
+                                PatternRewriter &rewriter) const override {
+    LDBG() << "Processing CoalescedGatherDMAOp: " << dmaOp;
+
+    auto forallOp = dmaOp->getParentOfType<scf::ForallOp>();
+    if (!forallOp) {
+      LDBG() << "FAILED - not inside scf.forall";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "coalesced_gather_dma not inside scf.forall");
+    }
+    LDBG() << "Found parent forall";
+
+    if (failed(verifyThreadMapping(forallOp))) {
+      LDBG() << "FAILED - thread mapping verification failed";
+      return rewriter.notifyMatchFailure(
+          dmaOp, "forall does not have proper thread mapping");
+    }
+    LDBG() << "Thread mapping verified";
+
+    if (failed(verifyMemoryLayout(dmaOp, rewriter))) {
+      LDBG() << "FAILED - memory layout verification failed";
+      return failure();
+    }
+    LDBG() << "Memory layout verified";
+
+    Value source = dmaOp.getSource();
+    Value dest = dmaOp.getInit();
+    auto indicesRange = dmaOp.getIndices();
+
+    auto sourceType = cast<MemRefType>(source.getType());
+    auto destType = cast<MemRefType>(dest.getType());
+
+    if (!indicesRange.empty()) {
+      Value indices = indicesRange.front();
+      LDBG() << "  Indices: " << indices;
+      auto indicesType = cast<MemRefType>(indices.getType());
+      ArrayRef<int64_t> indicesShape = indicesType.getShape();
+      ArrayRef<int64_t> destShape = destType.getShape();
+
+      // Verify that indices dimensions are a prefix of dest dimensions.
+      if (indicesShape.size() > destShape.size()) {
+        return rewriter.notifyMatchFailure(dmaOp,
+                                           "indices rank exceeds dest rank");
+      }
+
+      for (size_t i = 0; i < indicesShape.size(); ++i) {
+        if (indicesShape[i] != destShape[i]) {
+          return rewriter.notifyMatchFailure(
+              dmaOp, "indices shape is not a prefix of dest shape");
+        }
+      }
```

Comment:
1. Don't recalculate end
2. You can replace the whole loop with `if (indicesShape != destShape)`